### PR TITLE
fix(transformers): fix some bugs

### PR DESCRIPTION
**Pull request by Cui-yshoho**
# What does this PR do?

</p>
<hr>
<h1>SDPA Issues</h1>

### **Bug 1**

`mindspore.ops.speed_fusion_attention` is **not equivalent** to `torch.nn.functional.scaled_dot_product_attention`.  

Currently, PyTorch's `scaled_dot_product_attention` supports three implementations:

1. **[FlashAttention-2: Faster Attention with Better Parallelism and Work Partitioning](https://arxiv.org/abs/2307.08691)**
2. **[Memory-Efficient Attention](https://github.com/facebookresearch/xformers)**
3. A **C++ PyTorch implementation** matching the standard formulation.

> If the user requires a specific fused implementation, the C++ PyTorch implementation can be disabled via [`torch.nn.attention.sdpa_kernel()`](https://docs.pytorch.org/docs/stable/generated/torch.nn.attention.sdpa_kernel.html#torch.nn.attention.sdpa_kernel).

`ops.speed_fusion_attention` in MindSpore is used for **self-attention fusion computation**.

#### Scaled Dot Product Attention Implementation and NaN Handling

I implemented a `scaled_dot_product_attention` function **referencing `torch.nn.functional.scaled_dot_product_attention`**, with additional handling for causal masks.

##### Forward Pass

In my implementation, I explicitly check for **entire rows in the attention logits**. If a row consists entirely of `-inf` values before the softmax, the forward result would normally become `NaN`:

$$
\text{softmax}([-\infty, -\infty, \dots, -\infty]) 
= \frac{[e^{-\infty}, e^{-\infty}, \dots]}{\sum_i e^{-\infty}} 
= \frac{[0, 0, \dots]}{0} 
= \text{NaN}
$$

However, in `torch.nn.functional.scaled_dot_product_attention`, such rows are safely replaced with zeros:

$$
\text{softmax}([-\infty, -\infty, \dots, -\infty]) = [0, 0, \dots, 0]
$$

This prevents `NaN` in the **forward pass**.

##### Backward Pass

Without this handling, the backward gradients also become `NaN`. 
Mathematically, for softmax: $softmax(x)_i = e^{x_i} / \sum_j e^{x_j}$.

$$
\frac{\partial \text{softmax}(x)_i}{\partial x_k} =
\begin{cases}
\text{softmax}(x)_i (1 - \text{softmax}(x)_i), & \text{if } i = k \\
-\text{softmax}(x)_i \text{softmax}(x)_k, & \text{if } i \neq k
\end{cases}
$$

If `x = [-inf, -inf, ..., -inf]`, then:

$$
\text{softmax}(x)_i = \text{NaN}, \quad
\frac{\partial \text{softmax}(x)_i}{\partial x_k} = \text{NaN}.
$$

By replacing the entire row with zeros before and after softmax, both **forward and backward passes** are well-defined:

$$
softmax^{safe}(x)_i = 0, \quad 
\frac{\partial}{\partial x_k} softmax^{safe}(x)_i = 0
$$

<hr>
<h3><strong>Bug 2</strong></h3>
<p>When <code inline="">is_causal=True</code>, <code inline="">torch.nn.functional.scaled_dot_product_attention</code> <strong>does not</strong> simply assert <code inline="">attn_mask is None</code>.<br>
Instead:</p>
<ul>
<li>
<p>If <code inline="">attn_mask</code> is <strong>boolean</strong>, the result is the <strong>intersection</strong> of <code inline="">is_causal</code> and <code inline="">attn_mask</code>.</p>
</li>
<li>
<p>If <code inline="">attn_mask</code> is <strong>floating-point</strong>, the result is the <strong>addition</strong> of <code inline="">is_causal</code> and <code inline="">attn_mask</code>.</p>
</li>
</ul>
<p>This can be verified by the following examples:</p>

> **Note:** Boolean Mask

```python
import torch
import torch.nn.functional as F

batch_size, num_heads, seq_len, head_dim = 1, 1, 4, 8
query = torch.randn(batch_size, num_heads, seq_len, head_dim)
key = torch.randn(batch_size, num_heads, seq_len, head_dim)
value = torch.randn(batch_size, num_heads, seq_len, head_dim)

attention_mask = torch.ones(seq_len, seq_len, dtype=torch.bool)
attention_mask[2:, :] = False   # user-defined mask
causal_mask = torch.tril(torch.ones(seq_len, seq_len, dtype=torch.bool), 0)

# condition 1: is_causal=True + attention_mask
out1 = F.scaled_dot_product_attention(query, key, value,
                                      attn_mask=attention_mask,
                                      is_causal=True,
                                      dropout_p=0.0)

# condition 2: only attention_mask
out2 = F.scaled_dot_product_attention(query, key, value,
                                      attn_mask=attention_mask,
                                      is_causal=False,
                                      dropout_p=0.0)

# condition 3: only causal
out3 = F.scaled_dot_product_attention(query, key, value,
                                      attn_mask=None,
                                      is_causal=True,
                                      dropout_p=0.0)

# condition 4: attention_mask &amp; causal_mask
out4 = F.scaled_dot_product_attention(
    query, key, value,
    attn_mask= attention_mask & causal_mask,
    is_causal=False,
    dropout_p=0.0,
)

print(torch.allclose(out1, out2))  # False
print(torch.allclose(out1, out3))  # False
print(torch.allclose(out1, out4))  # True
```

> **Note:** Floating-Point Mask

```python
import torch
import torch.nn.functional as F

L = 4
query = torch.randn(1, 1, L, 8)
key   = torch.randn(1, 1, L, 8)
value = torch.randn(1, 1, L, 8)

attention_mask = torch.zeros(L, L)
attention_mask[2:, :] = float('-inf')

# causal mask
causal_mask = torch.triu(torch.full((L, L), float('-inf')), diagonal=1)

# condition 1: causal=True + attention_mask
out1 = F.scaled_dot_product_attention(query, key, value,
                                      attn_mask=attention_mask,
                                      is_causal=True)

# condition 2: causal=False + attention_mask + causal_mask
out2 = F.scaled_dot_product_attention(query, key, value,
                                      attn_mask=attention_mask + causal_mask,
                                      is_causal=False)

print(torch.allclose(out1, out2))  # True
```
<hr>
<h3><strong>Other Notes</strong></h3>
<ul>
<li>
<p><code inline="">ops.norm</code> does not support <code inline="">bf16</code>, replace it with <code inline="">mint.norm</code>.</p>
</li>
<li>
<p>Add <code inline="">sdpa</code> test cases.</p>
</li>
<li>
<p><code inline="">mint.nn.BatchNorm2d</code>:</p>
<ul>
<li>
<p>If we use <code inline="">set_dtype</code> on the <code inline="">Parameter</code>, then <code inline="">num_batches_tracked</code> (which should be <code inline="">int64</code>) will also be converted.</p>
</li>
<li>
<p>But in <code inline="">state_dict</code>, <code inline="">num_batches_tracked</code> is <code inline="">int64</code> in fact.</p>
</li>
<li>
<p>Solution: remove <code inline="">.to(ms.float32)</code> and add a proper check in <code inline="">modeling_test_utils.py</code>.</p>
</li>
</ul>
</li>
</ul>
<hr>
<p>

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/mindspore-lab/mindone/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the documentation with your changes? E.g. record bug fixes or new features in `What's New`. Here are the
      [documentation guidelines](https://github.com/mindspore-lab/mindcv/wiki/%E6%96%87%E6%A1%A3%E7%BC%96%E5%86%99%E8%A7%84%E8%8C%83)
- [ ] Did you build and run the code without any errors?
- [ ] Did you report the running environment (NPU type/MS version) and performance in the doc? (better record it for data loading, model inference, or training tasks)
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@xxx

### CODE REVIEW

**Comment by hadipash** [comment]
Good fixes! Interestingly, the PyTorch docs state that `An error is thrown if both attn_mask and is_causal are set` which is the opposite of the actual behavior.
https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#:~:text=An%20error%20is%20thrown%20if%20both%20attn_mask%20and%20is_causal%20are%20set

Please expand the test cases in `tests/transformers_tests/integrations/test_sdpa_attention.py` to cover your changes, as SDPA is crucial to the project.

**Review by vigo999 — APPROVED** [review]
